### PR TITLE
Replace Kotlin getters/setters with property access in codegen #496

### LINKED ISSUE #496: Compilation errors in code with setters generated on Kotlin
**Description**

Create Kotlin class

```kotlin
class A {
    var a: Int = 0;
}
```

Create Java class

```java
public class D {
    public boolean foo(A a) {
        return a.getA() > 0;
    }
}
```

and create tests for it. Assemble models are used in tests generation

**Expected behavior**

Generated tests code is compilable.

**Actual behavior**

The following code is generated

```kotlin
@Test
@DisplayName("foo: AGetA -> return a.getA() > 0")
fun testFoo_AGetAGreaterThanZero() {
    val d = D()
    val a = A()
    a.setA(1)
    
    val actual = d.foo(a)
    
    assertTrue(actual)
    }
```

**Visual proofs (screenshots, logs, images)**

image

### PR DESCRIPTION
**Pull request by volivan239**
# Description

Now codegen properly treats Kotlin's autogenerated getters/setters and uses direct property access instead of them.

Kotlin properties are treated as fields. However, instead of `FieldId.isAccessibleFrom` which is now private, methods `FieldId.canBeReadFrom` and `FieldId.canBeSetFrom` should be used. First one behaves for Java code in the same way as `isAccessibleFrom` did before, but for Kotlin it also returns true if there is an accessible getter. Behavior of `FieldId.canBeSetFrom` hasn't been changed for Java, but for Kotlin it now also returns true if there is accessible setter/getter pair. See [docs](https://kotlinlang.org/docs/java-interop.html#getters-and-setters) for more details.

These two methods guarantee that codegen produces code with field access where it is possible. However, some calls to getters/setters may be produced by `AssembleModelGenerator`, so this PR also adds `replaceCgExecutableCallWithFieldAccessIfNeeded` to replace such calls with direct accesses.

Fixes #496

## Type of Change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

## Manual Scenario 

Works correctly on examples from #496 and #948

# Checklist (remove irrelevant options):

_This is the author self-check list_

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] New documentation is provided or existed one is altered
- [x] No new warnings
- [ ] New tests have been added
- [ ] All tests pass locally with my changes

### CODE REVIEW

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt [hunk 1]
```diff
@@ -212,7 +217,12 @@ internal class CgVariableConstructor(val context: CgContext) :
                     instance[statementModel.fieldId] `=` declareOrGet(statementModel.fieldModel)
                 }
                 is UtExecutableCallModel -> {
-                    +createCgExecutableCallFromUtExecutableCall(statementModel)
+                    val call = createCgExecutableCallFromUtExecutableCall(statementModel)
+                    val callOrAccess: CgStatement = replaceCgExecutableCallWithFieldAccessIfNeeded(call)
+                    if (callOrAccess is CgExecutableCall)
```
Comment (EgorkaKulikov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt:222]: Improve this code fragment to make it more readable, please

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt [hunk 2]
```diff
@@ -236,6 +246,7 @@ internal class CgVariableConstructor(val context: CgContext) :
         val initExpr = if (isPrimitiveWrapperOrString(type)) {
             cgLiteralForWrapper(params)
         } else {
+            // TODO: if instantiation chain could be a setter call, we need to replace it in Kotlin
```
Comment (EgorkaKulikov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt:249]: We have an instantiation call, now instantiation chain right now. Anyway, I can't imagine a situation when the top-level instantiation call is a setter.

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt [hunk 3]
```diff
@@ -261,6 +272,23 @@ internal class CgVariableConstructor(val context: CgContext) :
         return cgCall
     }
 
+    private fun replaceCgExecutableCallWithFieldAccessIfNeeded(call: CgExecutableCall): CgStatement {
+        if (call !is CgMethodCall || context.codegenLanguage != CodegenLanguage.KOTLIN)
```
Comment (EgorkaKulikov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt:276]: For me it seems better to write code-specific logic in a way like
```
when (context.codegenlangauge) {
is Java -> return call
is Kotlin -> do some specific logic
}
```

But I do not insist on it.

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt [hunk 4]
```diff
@@ -261,6 +272,23 @@ internal class CgVariableConstructor(val context: CgContext) :
...
+            return call
+
+        val caller = call.caller ?: return call
+
+        caller.type.fieldThisIsSetterFor(call.executableId)?.let {
+            return CgAssignment(caller[it], call.arguments.single())
```
Comment (EgorkaKulikov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt:282]: Why don't you require that arguments contains single item here as in line 285?

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt [hunk 5]
```diff
@@ -261,6 +272,23 @@ internal class CgVariableConstructor(val context: CgContext) :
...
+        }
+        caller.type.fieldThisIsGetterFor(call.executableId)?.let {
+            require(call.arguments.isEmpty())
```
Comment (EgorkaKulikov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt:285]: require is more useful with a message with failure description

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/ClassIdUtil.kt [hunk 6]
```diff
@@ -28,4 +31,16 @@ infix fun ClassId.isAccessibleFrom(packageName: String): Boolean {
...
     }
-}
\ No newline at end of file
+}
+
+/**
+ * Returns field of [this], such that [methodId] is a getter for it (or null if methodId doesn't represent a getter)
+ */
+internal fun ClassId.fieldThisIsGetterFor(methodId: MethodId): FieldId? =
+    allDeclaredFieldIds.firstOrNull { !it.isStatic && it.getter == methodId }
```
Comment (EgorkaKulikov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/ClassIdUtil.kt:40]: May be `singleOrNull`? Same for the second method.

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/ClassIdUtil.kt [hunk 7]
```diff
@@ -28,4 +31,16 @@ infix fun ClassId.isAccessibleFrom(packageName: String): Boolean {
...
+
+/**
+ * Returns field of [this], such that [methodId] is a setter for it (or null if methodId doesn't represent a setter)
+ */
+internal fun ClassId.fieldThisIsSetterFor(methodId: MethodId): FieldId? =
```
Comment (EgorkaKulikov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/ClassIdUtil.kt:45]: I suggest the name `fieldThatIsSetWith`.
Current name means that field is a setter isself, it seems strange.

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/UtilMethods.kt [hunk 8]
```diff
@@ -343,7 +343,7 @@ private fun createInstance(visibility: Visibility, language: CodegenLanguage): S
         }
         CodegenLanguage.KOTLIN -> {
             """
-            ${visibility by language}fun createInstance(className: String): kotlin.Any? {
+            ${visibility by language}fun createInstance(className: String): kotlin.Any {
```
Comment (EgorkaKulikov) [inline utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/visitor/UtilMethods.kt:346]: To be honest, I don't understand this change, but can believe that is is possible :)

**Review by EgorkaKulikov — APPROVED** [review]

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt (hunk 1 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt]: Should be clearer now. Also, changed return type of `replaceCgExecutableCallWithFieldAccessIfNeeded` to be nullable and added docs to it

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt (hunk 2 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt]: Thanks for advice, deleted the question

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt (hunk 3 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt]: Fixed here, didn't check for other places

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt (hunk 4 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt]: I do -- at `call.arguments.single()`....

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt (hunk 5 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/constructor/tree/CgVariableConstructor.kt]: Added message

File: utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/ClassIdUtil.kt (hunk 7 above)
Comment (volivan239) [reply utbot-framework/src/main/kotlin/org/utbot/framework/codegen/model/util/ClassIdUtil.kt]: Renamed this and the method below

**Pull request merged**